zcash_primitives/src/sapling.rs

```diff
@@ -485,3 +512,58 @@ impl Note {
         }
     }
 }
+
+#[cfg(any(test, feature = "test-dependencies"))]
+pub mod testing {
+    use proptest::prelude::*;
+    use std::cmp::min;
+    use std::convert::TryFrom;
+
+    use crate::{
+        transaction::components::amount::MAX_MONEY, zip32::testing::arb_extended_spending_key,
+    };
+
+    use super::{Node, Note, NoteValue, PaymentAddress, Rseed};
+
+    prop_compose! {
+        pub fn arb_note_value()(value in 0u64..MAX_MONEY as u64) -> NoteValue {
+            NoteValue::try_from(value).unwrap()
+        }
+    }
+
+    prop_compose! {
+        /// The
+        pub fn arb_positive_note_value(bound: u64)(
+            value in 1u64..(min(bound, MAX_MONEY as u64))
+        ) -> NoteValue {
+            NoteValue::try_from(value).unwrap()
+        }
+    }
+
+    pub fn arb_payment_address() -> impl Strategy<Value = PaymentAddress> {
+        arb_extended_spending_key()
+            .prop_map(|sk| sk.default_address().map(|(_, a)| a))
+            .prop_filter("A valid payment address is required.", |r| r.is_ok())
+            .prop_map(|r| r.unwrap())
+    }
+
+    prop_compose! {
+        pub fn arb_node()(value in prop::array::uniform32(prop::num::u8::ANY)) -> Node {
+            Node::new(value)
+        }
+    }
+
+    prop_compose! {
+        pub fn arb_note(value: NoteValue)(
+            addr in arb_payment_address().prop_filter("Address must produce a valid g_d", |a| a.g_d().is_some()),
```

Comment:
The filter here is unnecessary, because you have already filtered for valid payment addresses inside `arb_payment_address()`. This is a deficiency of `PaymentAddress::g_d` (due to earlier non-typesafe constructability of `PaymentAddress`) that we should clean up at some point.
```suggestion
            addr in arb_payment_address(),
```